tech-spec.md

```diff
@@ -0,0 +1,361 @@
+# Tech spec
+
+The ODM sharing library aims to be able to read the most used file formats and
+databases, provide useful feedback to help build correct filters, and output
+easy-to-share result data.
+
+## Overview
+
+A sharing schema (in CSV format) is used to define individual rules (as rows),
+which come together to form a data query.
+
+The query rules are parsed into an abstract syntax tree (AST), which represents
+the structure of the query. It effectively separates rule-parsing from
+query-generation, making it more modular. The AST can then be used to generate
+concrete queries in SQL or other query languages, or it can be interpreted to
+perform data transformations directly using a library like Pandas.
+
+The library will focus on using SQL as the query language, since it has wide
+support and is easy to generate. Using SQL will also provide a free performance
+boost when running queries on big indexed databases, as well as enable us to
+output the intermediate SQL to the user in case they want to inspect or execute
+the queries themselves.
+
+## CLI
+
+**Usage**
+
+```bash
+./share.py [--orgs=<name[,...]>] [--outdir=<path>] <schema> <input>
+```
+
+- orgs: comma separated list of the organizations to output data for, or all if
+  not specified
+- outdir: output file directory
+- schema: sharing schema file path
+- input: spreadsheet file path or [SQLAlchemy database url](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls)
+
+**Examples**
+
+Create a collection of sharable csv files in the "~/ohri" directory, for the
+"OHRI" organization, applying the rules from schema.csv on the input from
+data.xlsx:
+
+```bash
+./share.py --orgs=OHRI --outdir=~/ohri/ schema.csv data.xlsx
+```
+
+Output to the default (current) directory, for all organizations specified in
```

Comment:
I assumed that the sqlalchemy would be doing everything?

Reply:
I'm assuming that the output is the same for a database as for an Excel file?

Reply:
yes, the user can specify either a file-path or a database-url as input, and the rest will be identical.

Reply:
yes, since python needs a way to communicate with the database, the user needs to install the appropriate python package for it (either a specific one like pymysql or a general one like pyodbc). Excel users won't need to install anything extra.

Reply:
done